enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1104 @@
+	// +kubebuilder:validation:Required
+	Selector metav1.LabelSelector `json:"selector"`
+
+	// Template describes the Control Plane Machines that will be created
+	// by this ControlPlaneMachineSet.
+	// +kubebuilder:validation:Required
+	Template ControlPlaneMachineSetTemplate `json:"template"`
+}
+
+// ControlPlaneMachineSetTemplate is a template used by the ControlPlaneMachineSet
+// to create the Machines that it will manage in the future.
+// +union
+// + ---
+// + This struct is a discriminated union which allows users to select the type of Machine
+// + that the ControlPlaneMachineSet should create and manage.
+// + For now, the only supported type is the OpenShift Machine API Machine, but in the future
+// + we plan to expand this to allow other Machine types such as Cluster API Machines or a
+// + future version of the Machine API Machine.
+type ControlPlaneMachineSetTemplate struct {
+  // MachineType determines the type of Machines that should be managed by the ControlPlaneMachineSet.
+	// Currently, the only valid value is machines_v1beta1_machine_openshift_io.
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	MachineType ControlPlaneMachineSetMachineType `json:"machineType"`
+
+	// OpenShiftMachineV1Beta1Machine defines the template for creating Machines
+	// from the v1beta1.machine.openshift.io API group.
+	// +kubebuilder:validation:Required
+	OpenShiftMachineV1Beta1Machine *OpenShiftMachineV1Beta1MachineTemplate `json:"machines_v1beta1_machine_openshift_io,omitempty"`
+}
+
+// ControlPlaneMachineSetMachineType is a enumeration of valid Machine types
+// supported by the ControlPlaneMachineSet.
+// +kubebuilder:validation:Enum:=machines_v1beta1_machine_openshift_io
+type ControlPlaneMachineSetMachineType string
+
+const (
+	// OpenShiftMachineV1Beta1MachineType is the OpenShift Machine API v1beta1 Machine type.
+	OpenShiftMachineV1Beta1MachineType ControlPlaneMachineSetMachineType = "machines_v1beta1_machine_openshift_io"
+)
+
+// OpenShiftMachineV1Beta1MachineTemplate is a template for the ControlPlaneMachineSet to create
+// Machines from the v1beta1.machine.openshift.io API group.
+type OpenShiftMachineV1Beta1MachineTemplate struct {
+	// FailureDomains is the list of failure domains (sometimes called
+	// availability zones) in which the ControlPlaneMachineSet should balance
+	// the Control Plane Machines.
+	// This will be merged into the ProviderSpec given in the template.
+	// This field is optional on platforms that do not require placement
+	// information, eg OpenStack.
+	// +optional
+	FailureDomains FailureDomains `json:"failureDomains,omitempty"`
+
+	// ObjectMeta is the standard object metadata
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// Labels are required to match the ControlPlaneMachineSet selector.
+	// +kubebuilder:validation:Required
+	ObjectMeta ControlPlaneMachineSetTemplateObjectMeta `json:"metadata"`
```

Comment:
As a rule in OpenShift APIs we avoid reusing upstream objects in our specs as they add fields to our APIs that we don't care about. For example, having the whole `metav1.ObjectMeta` here would mean fields like the name or creation timestamp are now a part of this API, yet they have no meaning to the controller and if a user sets them, are ignored.

It's better to have a struct with only those fields we actually care about so that the generated schema can give better feedback to the user about what is and isn't supported.